drivers/dma/dma_dw_axi.c

```diff
@@ -0,0 +1,1002 @@
+			" max descriptors in pool: %d", dev->name, channel,
+			cfg->block_count, CONFIG_DMA_DW_AXI_MAX_DESC);
+		return -EINVAL;
+	}
+
+	if (cfg->source_burst_length > CONFIG_DMA_DW_AXI_MAX_BURST_TXN_LEN ||
+			cfg->dest_burst_length > CONFIG_DMA_DW_AXI_MAX_BURST_TXN_LEN ||
+			cfg->source_burst_length == 0 || cfg->dest_burst_length == 0) {
+		LOG_ERR("dma:%s burst length not supported", dev->name);
+		return -ENOTSUP;
+	}
+
+	/* get channel specific data pointer */
+	chan_data = &dw_dev_data->chan[channel - 1];
+
+	k_mutex_lock(&chan_data->ch_mutex_lock, K_FOREVER);
+
+	/* check if the channel is currently idle */
+	if (atomic_get(&chan_data->ch_state) != DMA_DW_AXI_CH_IDLE) {
+		k_mutex_unlock(&chan_data->ch_mutex_lock);
+		LOG_ERR("DMA channel:%d is busy", channel);
+		return -EBUSY;
+	}
+
+	/* burst transaction length for source and destination */
+	msize_src = DMA_DW_AXI_GET_MSIZE(cfg->source_burst_length);
+	msize_dst = DMA_DW_AXI_GET_MSIZE(cfg->dest_burst_length);
+
+	chan_data->cfg = 0;
+	chan_data->irq_unmask = 0;
+
+	chan_data->direction = cfg->channel_direction;
+
+	chan_data->lli_desc_base =
+			&dw_dev_data->dma_desc_pool[(channel - 1) * CONFIG_DMA_DW_AXI_MAX_DESC];
+	chan_data->lli_desc_count = cfg->block_count;
+	memset(chan_data->lli_desc_base, 0,
+			sizeof(struct dma_lli) * chan_data->lli_desc_count);
+
+	lli_desc = chan_data->lli_desc_base;
+	blk_cfg = cfg->head_block;
+
+	/* max channel priority can be MAX_CHANNEL - 1 */
+	if (cfg->channel_priority < dw_dma_config->max_channel) {
+		chan_data->cfg |= DMA_DW_AXI_CFG_PRIORITY(cfg->channel_priority);
+	}
+
+	/* configure all the descriptors in a loop */
+	for (i = 0; i < cfg->block_count; i++) {
+
+		ret = dma_dw_axi_set_data_width(lli_desc, cfg->source_data_size,
+				cfg->dest_data_size);
+		if (ret) {
+			k_mutex_unlock(&chan_data->ch_mutex_lock);
+			return ret;
+		}
+
+		lli_desc->ctl |= DMA_DW_AXI_CTL_SRC_STAT_EN |
+				DMA_DW_AXI_CTL_DST_STAT_EN | DMA_DW_AXI_CTL_IOC_BLK_TFR;
+
+		lli_desc->sar = blk_cfg->source_address;
+		lli_desc->dar = blk_cfg->dest_address;
+
+		/* set block transfer size*/
+		lli_desc->block_ts_lo = (blk_cfg->block_size / cfg->source_data_size) - 1;
+		if (lli_desc->block_ts_lo > CONFIG_DMA_DW_AXI_MAX_BLOCK_TS) {
+			k_mutex_unlock(&chan_data->ch_mutex_lock);
+			LOG_ERR("block transfer size more than %u not supported",
+				CONFIG_DMA_DW_AXI_MAX_BLOCK_TS);
+			return -ENOTSUP;
+		}
+
+		/* configuration based on channel direction */
+		if (cfg->channel_direction == MEMORY_TO_MEMORY) {
+			chan_data->cfg |= DMA_DW_AXI_CFG_TT_FC(M2M_DMAC);
+
+			lli_desc->ctl |= DMA_DW_AXI_CTL_SRC_MSIZE(msize_src) |
+					DMA_DW_AXI_CTL_DST_MSIZE(msize_dst);
+
+		} else if (cfg->channel_direction == MEMORY_TO_PERIPHERAL) {
+
+			chan_data->cfg |= DMA_DW_AXI_CFG_TT_FC(M2P_DMAC);
+			lli_desc->ctl |= DMA_DW_AXI_CTL_SRC_MSIZE(msize_src) |
+					DMA_DW_AXI_CTL_DST_MSIZE(msize_dst);
+			WRITE_BIT(chan_data->cfg, DMA_DW_AXI_CFG_HW_HS_DST_BIT_POS, 0);
```

Comment:
@carlocaione
We can use sys_clear_bit but, I think sys_* functions are mainly used to write onto the memory mapped registers as internally it will perform a volatile read or write.

In this case, chan_data->cfg is a global variable, so I thought it would be better to use WRITE_BIT